ui/wallet_page.go

```diff
@@ -0,0 +1,173 @@
+		editorW:       common.theme.Editor("Enter wallet name"),
+		addAcctW:      common.theme.IconButton(common.icons.contentAdd),
+		deleteW:       common.theme.DangerButton("Delete Wallet"),
+	}
+	gtx := common.gtx
+
+	return func() {
+		current := common.info.Wallets[*common.selectedWallet]
+		if common.walletsTab.Changed() {
+			page.renaming = false
+		}
+		wdgs := []func(){
+			func() {
+				if page.renaming {
+					horFlex.Layout(gtx,
+						rigid(func() {
+							page.editorW.Layout(gtx, &page.editor)
+						}),
+						rigid(func() {
+							page.renameW.Layout(gtx, &page.rename)
+						}),
+						rigid(func() {
+							layout.Center.Layout(gtx, func() {
+								page.cancelRenameW.Layout(gtx, &page.cancelRename)
+							})
+						}),
+					)
+				} else {
+					horFlex.Layout(gtx,
+						rigid(func() {
+							common.theme.H1(current.Name).Layout(common.gtx)
+						}),
+						rigid(func() {
+							layout.Center.Layout(gtx, func() {
+								page.beginRenameW.Layout(gtx, &page.beginRename)
+							})
+						}),
+					)
+				}
+			},
+			func() {
+				common.theme.H5("Total Balance: " + current.Balance).Layout(gtx)
+			},
+			func() {
+				horFlex.Layout(gtx,
+					rigid(func() {
+						common.theme.H5("Accounts").Layout(gtx)
+					}),
+					rigid(func() {
+						layout.S.Layout(gtx, func() {
+							page.addAcctW.Layout(gtx, &page.addAcct)
+						})
+					}),
+				)
+			},
+			func() {
+				page.accountsList.Layout(gtx, len(current.Accounts), func(i int) {
+					acct := current.Accounts[i]
+					a := func() {
+						layout.Flex{Axis: layout.Vertical}.Layout(gtx,
+							layout.Rigid(func() {
+								common.theme.Body1(acct.Name).Layout(gtx)
+							}),
+							layout.Rigid(func() {
+								common.theme.Body1(acct.TotalBalance).Layout(gtx)
+							}),
+							layout.Rigid(func() {
+								common.theme.Body1("Keys: " + acct.Keys.External + " external, " + acct.Keys.Internal + " internal, " + acct.Keys.Imported + " imported").Layout(gtx)
+							}),
+							layout.Rigid(func() {
+								common.theme.Body1("HD Path: " + acct.HDPath).Layout(gtx)
+							}),
+						)
+					}
+					layout.UniformInset(unit.Dp(5)).Layout(gtx, a)
+				})
+			},
+			func() {
+				page.deleteW.Layout(gtx, &page.delete)
+			},
+		}
+
+		page.dialog.LayoutIfActive(gtx, func() {
+			common.LayoutWithWallets(gtx, func() {
+				page.container.Layout(common.gtx, len(wdgs), func(i int) {
+					wdgs[i]()
+				})
+			})
+		})
+
+		if page.beginRename.Clicked(gtx) {
+			page.renaming = true
+			page.editor.SetText(current.Name)
+		}
+
+		if page.cancelRename.Clicked(gtx) {
+			page.renaming = false
+		}
+
+		if page.rename.Clicked(gtx) {
+			name := page.editor.Text()
+			err := common.wallet.RenameWallet(current.ID, name)
+			if err != nil {
+				log.Error(err)
+			} else {
+				common.info.Wallets[*common.selectedWallet].Name = name
+				page.renaming = false
+			}
+		}
+
+		if page.delete.Clicked(gtx) {
+			page.dialog.SetDialog(page.dialog.deleteWallet(common))
+		}
```

Comment:
The event handling conditions should be in a separate method, the result is a much cleaner code. To make sure every page has an event handling method, you can define an interface which every valid page must implement.

Reply:
I actually prefer this to a separate method given that the only requirement a page has to satisfy now is returning a `layout.Widget` so adding an interface seems like unnecessary complexity.